### ISSUE TITLE
UnhandledPromiseRejectionWarning: Error: Unsupported image type

### ISSUE THREAD

**Issue by Devidian** [Automattic/node-canvas#1481]
Hi, in one of my project there is an error occurring when the image provided is not an image (user/customer input). It results in an unhandled promise rejection. Sure i have to check content before i pass it to image.src but there could always be a case something is breaking through.

## Issue or Feature
```log
(node:21275) UnhandledPromiseRejectionWarning: Error: Unsupported image type
     at setSource (./node_modules/canvas/lib/image.js:91:13)
     at Image.set (./node_modules/canvas/lib/image.js:65:7)
     at ./bin/ImageLoader.js:53:25
     ... 
```

## Steps to Reproduce

```ts
return new Promise((resolve, reject) => {
	let img = new Image();
	let buf: Buffer = null;
	img.onload = () => {
		const canvas = createCanvas(img.width, img.height);
		canvas.getContext("2d").drawImage(img, 0, 0);
		resolve(canvas);
	};
	img.onerror = (e: any) => {
		Logger(911, "ImageLoader->loadImage->img.onerror", `Error loading image from path ${sourcePath}`);
		reject(e);
	};
	if (!this.cache.has(index)) {
		try {
			accessSync(sourcePath);
			buf = readFileSync(sourcePath); // should be image but could be broken image or other file like xml without checks
			this.cache.set(index, buf);
		} catch (error) {
			Logger(911, "ImageLoader->loadImage->readFileSync", `Error reading image on path ${sourcePath}`);
			reject(error);
			return;
		}
	} else {
		buf = this.cache.get(index);
	}
	try {
		img.src = buf; // This is "./bin/ImageLoader.js:53:25"
	} catch (error) {
		reject(error);
	}
});
```

## Your Environment
* Version of node-canvas (output of `npm list canvas` or `yarn list canvas`): canvas@2.6.0
* Environment (e.g. node 4.2.0 on Mac OS X 10.8): Debian GNU/Linux 9 \n \l

**Comment by LinusU** [comment]
Well.... are you handling the Promise rejection? 😄 

It's your code that are constructing the `Promise`, and subsequently rejecting it in `./bin/ImageLoader.js` on line `55`...

**Comment by Devidian** [comment]
Hm sure,... just checked the complete call its a bit more complex, here just the relevant parts. Maybe i just cant see the gap thats missing...
```ts


class Renderer {
    protected async animate(frame: number): Promise<boolean> {
        let animationPromise = Promise.resolve(true);
        // ...
        animationPromise.then(() => {
            return this.blockAnimation(frame, bc, index); // <==== 1
        }).catch((e) => { // should be catched here
            console.log(LOGTAG.ERROR, '[Renderer.animate]', e.message);
            !cfg.log.debug ? null : console.log(LOGTAG.DEBUG, '[Renderer.animate]', e);
        });
        // ...
    }

    protected async blockAnimation(frame: number, blockConfig: BlockConfig, itemIndex: number): Promise<boolean> {
        try {
            // ...
            const ItemImage = await this.ImageLoader.getProductImage(Item, 'imageUrl'); // <==== 2
            // ...
        } catch (error) {
            Logger(911, "HD2-Jumbo", error);
			ctx.restore();
			throw error;
        }
    }
}

class ImageLoader{
    public async getProductImage(Product: Product, prop: string): Promise<Canvas> {
		if (!Product) return this.noImage;
		let localItemImagePath = this.getProductCacheLocalPath(Product, prop);

		try {
			accessSync(localItemImagePath);
            const canvas = await this.loadCachedImage(localItemImagePath, localItemImagePath);  // <==== 3
            return canvas;
		} catch (error) {
            console.log(LOGTAG.ERROR, '[ImageLoader.getProductImage]', error.message);
            const canvas = await this.downloadImageFile(Product[prop && Product[prop] ? prop : 'imageUrl'], localItemImagePath);
            return canvas;
		}
    }
    
    public async loadCachedImage(index: string, sourcePath: string): Promise<Canvas> {
		return new Promise((resolve, reject) => {
			let img = new Image();
			let buf: Buffer = null;
			img.onload = () => {
				const canvas = createCanvas(img.width, img.height);
				canvas.getContext("2d").drawImage(img, 0, 0);
				resolve(canvas);
			};
			img.onerror = (e: any) => {
				Logger(911, "ImageLoader->loadImage->img.onerror", `Error loading image from path ${sourcePath}`);
				reject(e); // <= this should start rejection chain
			};
			if (!this.cache.has(index)) {
				try {
					accessSync(sourcePath);
					buf = readFileSync(sourcePath); // should be image but could be broken image or other file like xml without checks
					this.cache.set(index, buf);
				} catch (error) {
					Logger(911, "ImageLoader->loadImage->readFileSync", `Error reading image on path ${sourcePath}`);
					reject(error);
					return;
				}
			} else {
				buf = this.cache.get(index);
			}
			try {
				img.src = buf; // This is "./bin/ImageLoader.js:53:25" // <==== 4
			} catch (error) {
				reject(error);
			}
		});
	}
}
```

**Comment by zbjornson** [comment]
This doesn't appear to be an error in node-canvas, as @LinusU pointed out. It's hard to follow your code snippet, but the stack trace should point you in the right direction of your unhandled rejection.
